Feature: Time In Service

pheonixstorm opened the pull request:
Allows recruitment date tracking based on hire date as well as freed
prisoner/bondsman and hired dependents.

@Akjosch  If you get a change look over this and make sure I didn't miss anything. I did not run into any bugs and the code looks fine. AFAIK it is ready to be merged.

Akjosch commented:
One thing I'd change: Fill the `recruitment` variable with proper data regardless of the options. You can change the campaign options mid-game, after all.

pheonixstorm commented:
If the option is off recruitment will be null and ignored in other areas like the majority of other options.

It's one of the reason I put various checks for both the option and the null value. That null does have its uses.

Akjosch commented:
It can and should still be ignored. The date should still be *recorded*, so that you don't have to guess when you decide to turn on the option later.

pheonixstorm commented:
You could say the same for the AtB init as well though which is what the time in service init is based on. Find the first log entry, check for a joined string. If not (rare) take a guess. Out of all of the campaigns I have looked over off the SF and GitHub trackers I have yet to run across one that has had log entries removed. Which was the main reason I started the log saver patches.

Akjosch commented:
Well, I would say we *don't* take how the AtB implementation does things as something to be emulated.

The issue here is the balance of the (memory) overhead of having unused data vs. the (processing) overhead of regenerating this data when we start to use it anyway. And in this case, the regenerating overhead loses on the account of not being able to recreate identical results to simply storing the data in the first place.

Now, why MekHQ uses `GeorgianCalendar` with its "let's waste more than 400 bytes to store a point in time" approach instead of ... I don't know, maybe a wrapped `long` of seconds since 1900 is another question and if you used `java.util.Date` or simply `long` to store that, I wouldn't mind.

pheonixstorm commented:
400 bytes to store so few characters... Well, that makes 3 I can easily think of. Birthday, deathday, and recruitment date. Maybe the log date too.

As for identical results. Weather we take in the data before hand or whenever the option is turned on the likelihood of the data not being identical is less than 1%

Another item I thought of is that prisoners do not get a recruitment flag, nor dependents due to the flag flip. Not to mention neither of them ever get a joined date anyway, so if you preload the recruitment info then turn the option on later dependents and prisoners will always be null until you edit them and use the date chooser in the customize dialog.

With a long, Date, or JodaTime how easy is it to covert to give years, months, hours, etc. One thing I will need eventually is a way to handle the number of months a given person has spent at a certain rank. Right now I haven't looked into it use calendar but... the others are, I think, completely foreign to me.

Akjosch commented:
JodaTime is relatively easy, though you have to get your head around Instant, DateTime (which is a form of Instant with attached *meaning* to it), Interval, Duration and Period and what they mean.

Though essentially, given two Instant or DateTime instances:

    int years = new Period(then, now, PeriodType.years()).getYears();
    int months = new Period(then, now, PeriodType.months()).getMonths();
    int hours = new Period(then, now, PeriodType.hours()).getHours();

pheonixstorm commented:
Looks like it might be easy to use. So the only thing I can think of is how it would interact with the date chooser and how much would need to change to display the values.

Another issue would be backwards compatibility.

Akjosch commented:
For "Time in Service", you don't need to worry about backwards compatibility, since there is none. The others can be changed later.

I'd just like to have the `recruitment` field for everyone set, saved and loaded (though not necessarily initialised if it's not there) regardless of the option setting.

pheonixstorm commented:
Backwards compatibility for the regular save games if gregoriancalendar is changed. But.. on with my main question.

I still see no real benefit to preloading the data over waiting for the option to be turned on. How do we do it? Part of the code to do so is already there but you still have to run the init whenever a campaign gets loaded just to make sure the data is there which is really a waste of time (albeit a tiny one).

Having this run in the background on a new campaign is simple and the code is already there. The real question is how to handle things for an existing campaign w/o having to do it repeatedly whenever the same campaign is loaded.

It would be far easier to set this to a default value of true then let the player decide if they want to use it or not. That I can live with and makes more sense.

Akjosch commented:
My suggestion would be to simply remove the guards on setRecruitment(...).

There are three of those in `MekHQ/src/mekhq/campaign/Campaign.java`, one each in `MekHQ/src/mekhq/gui/dialog/HireBulkPersonnelDialog.java`, `MekHQ/src/mekhq/gui/dialog/NewRecruitDialog.java`, and two in `MekHQ/src/mekhq/gui/dialog/PersonnelMarketDialog.java
`. There's also the check in `MekHQ/src/mekhq/gui/dialog/CustomizePersonDialog.java`, inside `initializePilotAndOptions()`. All of those can go.

The method `setDependent()` in `MekHQ/src/mekhq/campaign/personnel/Person.java` *already* ignores the setting, so that's fine.

And `initTimeInService()` can be called as it is now, whenever that option gets turned on. It just needs to skip over everyone who already has a recruitment date set.

pheonixstorm commented:
Hence default this option to true (on) and wait for the user to turn it off. Satisfies your desires and mine w/o adding bloat to the code.

To preload the data requires init to be run for new campaigns (already have the code in place) as well as loading existing campaigns. It is loading for existing campaigns that makes the least sense.

Akjosch commented:
I'm ... not sure what you mean by "adding bloat to code".

My suggestion is to *remove* a total of 16 new lines from your PR and add three new. Essentially, turn every line like this:

	if (campaign.getCampaignOptions().getUseTimeInService()) {
	    GregorianCalendar rawrecruit = (GregorianCalendar) campaign.getCalendar().clone();
	    selectedPerson.setRecruitment(rawrecruit);
	}

Into:

    selectedPerson.setRecruitment(rawrecruit);

... and add a simple check for `initializePilotAndOptions()` at the beginning of the loop:

    if(null != p.getRecruitment()) {
        continue;
    }

That's like the *opposite* of "adding bloat" ...

pheonixstorm commented:
No the extra bloat comes in as adding extra code to add the recruitment dates to an existing save games rather than a new campaign. Though after thinking about it I would have to add the code anyway if the option was defaulted to on.

But by loading all of the information regardless of setting is there really any point any more of this being optional?? This adds a single line to the person panel but changes nothing in how HQ actually works.

Akjosch commented:
I don't really have an opinion on whether this should be optional or always-on. There are reasons for optionality (not cluttering the UI with info the player doesn't want, explicitly saying you don't want any campaign rules involving time in service and so on), but since every campaign rule I care about (mainly Mercenary's Handbook 3055 and MechWarrior 3) doesn't involve them, I'll defer *that* decision to all the people who care to comment.

pheonixstorm commented:
It's not really a rule but extra information about the particular person just like the UI Spouse addition, though in the case of time in service the information must be generated rather than pull from am option that is already used.

pheonixstorm commented:
@Akjosch 
This should do. Tracks what it needs to but does not show the info unless the option is turned on.

Akjosch commented:
I was actually more worried about the prisoner handling, since they don't get a "Joined ..." personnel log line.

pheonixstorm commented:
That was the main reason I added `// If by some nightmare there is no Joined date just use the first entry.` 

Also why I have a prisoner related log saver in the works. pheonixstorm/MekHQ-storm#17
I plan on expanding that to include the freed section and change it to defected and joined, and in turn update the TiS code to include defected as a keyword. Or at least that's the plan.

But... went ahead and added freed to the keyword search for now.

Akjosch commented:
You're making this way more complicated than it has to be.

As soon as somebody writes a localisation file, your checks for "Joined ..." and "Freed" will stop working. If you load those strings from the property bundle instead, they'll stop working as soon as someone loads the campaign file created in a different language environment than the one they are using.

Just make it *simpler*: Record the recruited date *always*, don't change the date for personnel which already has it unless it's from the "Edit Person" GM dialogue, don't sweat it if you need to guess which date to take when somebody doesn't have one - either just add the first personnel log date or display "unknown".

Akjosch commented:
Just to show a small amount of variations for where the "joined" in "joined the unit" ends up when you translate it:

Japanese: End of the text. ("ユニットに**参加しました**")

German: Either end of text ("Der Einheit **beigetreten**") or split up between beginning and end ("**Trat** der Einheit **bei**")

Polish: Beginning of text, but different for male ("**Przyłączył się** do jednostki") and female ("**Przyłączyła się** do jednostki")

pheonixstorm commented:
The entire log structure would have to be redesigned first. At present the entire thing is made up of multiple strings put together. Changes to how the log messages are put together would require a rather large rewrite and not something that would happen any time soon.

As such I have no plans to alter the design of this PR and have reset the code back to its original state.
